compiler/rustc_hir_analysis/src/delegation.rs

```diff
@@ -157,18 +203,71 @@ impl<'tcx> PredicatesBuilder<'tcx> {
         args: ty::GenericArgsRef<'tcx>,
         sig_id: DefId,
     ) -> PredicatesBuilder<'tcx> {
-        PredicatesBuilder { tcx, args, parent: None, sig_id }
+        PredicatesBuilder {
+            tcx,
+            sig_id,
+            parent: None,
+            inh_kind: InheritanceKind::Eagerly(true),
+            args,
+        }
+    }
+
+    fn with_parent(mut self, parent: DefId) -> Self {
+        self.parent = Some(parent);
+        self
+    }
+
+    fn with_inheritance_kind(mut self, inh_kind: InheritanceKind) -> Self {
+        self.inh_kind = inh_kind;
+        self
     }
 
     fn build(self) -> ty::GenericPredicates<'tcx> {
-        let mut preds = vec![];
+        struct PredicatesCollector<'tcx> {
+            tcx: TyCtxt<'tcx>,
+            preds: Vec<(ty::Clause<'tcx>, Span)>,
+            args: ty::GenericArgsRef<'tcx>,
+        }
+
+        impl<'tcx> PredicatesCollector<'tcx> {
+            fn new(tcx: TyCtxt<'tcx>, args: ty::GenericArgsRef<'tcx>) -> PredicatesCollector<'tcx> {
+                PredicatesCollector { tcx, preds: vec![], args }
+            }
 
-        let sig_predicates = self.tcx.predicates_of(self.sig_id);
-        if let Some(parent) = sig_predicates.parent {
-            let sig_parent_preds = self.tcx.predicates_of(parent);
-            preds.extend(sig_parent_preds.instantiate_own(self.tcx, self.args));
+            fn with_own_preds<F>(mut self, f: F, def_id: DefId) -> Self
+            where
+                F: Fn(DefId) -> ty::GenericPredicates<'tcx>,
+            {
+                let preds = f(def_id).instantiate_own(self.tcx, self.args);
+                self.preds.extend(preds);
+                self
+            }
+
+            fn with_preds<F>(mut self, f: F, def_id: DefId) -> Self
+            where
+                F: Fn(DefId) -> ty::GenericPredicates<'tcx> + Copy,
+            {
+                let preds = f(def_id);
+                if let Some(parent_def_id) = preds.parent {
+                    self = self.with_own_preds(f, parent_def_id);
+                }
+                self.with_own_preds(f, def_id)
+            }
+        }
+        let collector = PredicatesCollector::new(self.tcx, self.args);
+
+        let preds = match self.inh_kind {
+            InheritanceKind::Eagerly(false) => {
+                collector.with_preds(|def_id| self.tcx.explicit_predicates_of(def_id), self.sig_id)
+            }
+            InheritanceKind::Eagerly(true) => {
+                collector.with_preds(|def_id| self.tcx.predicates_of(def_id), self.sig_id)
```

Comment:
Why does the "copy `Self`" flag triggers the difference between `predicates_of` and `explicit_predicates_of`?
From what I understand `predicates_of` adds more than just `Self::Trait`.

Reply:
> From what I understand predicates_of adds more than just Self::Trait.

 `predicates_of` = `explicit_predicates_of` + `Self: Trait` + `inferred_outlives_of`.  For fn's and traits there are no  inferred outlives predicates.